File: 21. Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = ListNode(0)
        p = head
        i,j = l1, l2
        while i and j:
            if i.val < j.val:
                p.next = i
                p = i
                i = i.next
            else:
                p.next = j
                p = j
                j = j.next
        while i:
            p.next = i
            p = i
            i = i.next
        while j:
            p.next = j
            p = j
            j = j.next
        
        t = head
        ret = head.next
        del t
        return ret
            
            
            
        

File: go/src/leetcode/21. Merge Two Sorted Lists/21. Merge Two Sorted Lists.go
package main

type ListNode struct {
	Val int
	Next *ListNode
}

func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := new(ListNode)
	p := dummy
	for l1 != nil || l2 != nil {
		if l1 != nil && l2 != nil {
			if l1.Val < l2.Val {
				p.Next = l1
				l1 = l1.Next
			} else {
				p.Next = l2
				l2 = l2.Next
			}
		} else if l1 != nil {
			p.Next = l1
			l1 = l1.Next
		} else {
			p.Next = l2
			l2 = l2.Next
		}
		p = p.Next
	}

	return dummy.Next
}
